File: go-cert-tools/verify-cert.go
package main

import (
	"crypto/ecdsa"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"log"
	"os"
	"path/filepath"
)


func main() {
	// 引数が渡されているか確認
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run verify-cert.go <DOMAIN>")
		return
	}

	// ドメイン名を取得
	domain := os.Args[1]
	fmt.Printf("Verifying certificate for domain: %s\n", domain)

	// fqdn を宣言して値を代入
	fqdn := domain
	localPath := "../certificates/"

	// ファイルパスを指定
	certFile := filepath.Join(localPath, fqdn, "cert.pem")  // サーバー証明書
	chainFile := filepath.Join(localPath, fqdn, "chain.pem") // 中間証明書
	keyFile := filepath.Join(localPath, fqdn, "privkey.pem") // サーバー証明書の秘密鍵

	// サーバー証明書を読み込む
	serverCert := loadCertificate(certFile)
	// 中間証明書を読み込む
	intermediateCert := loadCertificate(chainFile)
	// 秘密鍵を読み込む
	privateKey := loadPrivateKey(keyFile)

	// サーバー証明書の詳細を表示
	displayServerDetails(serverCert)
	// 中間証明書の詳細を表示
	displayIntermediateDetails(intermediateCert)


	// サーバー証明書と中間証明書の整合性を確認
	err := serverCert.CheckSignatureFrom(intermediateCert)
	if err != nil {
			log.Fatalf("サーバー証明書と中間証明書の整合性が確認できません: %v", err)
	}
	fmt.Println("サーバー証明書と中間証明書の整合性が確認されました")

	// サーバー証明書と秘密鍵の整合性を確認
	switch key := privateKey.(type) {
	case *rsa.PrivateKey:
			// RSAの場合
			if rsaPub, ok := serverCert.PublicKey.(*rsa.PublicKey); ok {
					if rsaPub.N.Cmp(key.PublicKey.N) != 0 {
							log.Fatalf("サーバー証明書とRSA秘密鍵の整合性が確認できません")
					}
			} else {
					log.Fatalf("サーバー証明書の公開鍵がRSA形式ではありません")
			}
	case *ecdsa.PrivateKey:
			// ECDSAの場合
			if ecdsaPub, ok := serverCert.PublicKey.(*ecdsa.PublicKey); ok {
					if ecdsaPub.X.Cmp(key.PublicKey.X) != 0 || ecdsaPub.Y.Cmp(key.PublicKey.Y) != 0 {
							log.Fatalf("サーバー証明書とECDSA秘密鍵の整合性が確認できません")
					}
			} else {
					log.Fatalf("サーバー証明書の公開鍵がECDSA形式ではありません")
			}
	default:
			log.Fatalf("未知の秘密鍵形式です")
	}

	fmt.Println("サーバー証明書と秘密鍵の整合性が確認されました")
}

// 証明書を読み込む関数
func loadCertificate(filePath string) *x509.Certificate {
	data, err := os.ReadFile(filePath)
	if err != nil {
			log.Fatalf("ファイルの読み込みに失敗しました: %v", err)
	}

	block, _ := pem.Decode(data)
	if block == nil {
			log.Fatalf("PEMデータのデコードに失敗しました: %s", filePath)
	}

	cert, err := x509.ParseCertificate(block.Bytes)
	if err != nil {
			log.Fatalf("証明書のパースに失敗しました: %v", err)
	}

	return cert
}

// 秘密鍵を読み込む関数
func loadPrivateKey(filePath string) interface{} {
	data, err := os.ReadFile(filePath)
	if err != nil {
			log.Fatalf("ファイルの読み込みに失敗しました: %v", err)
	}

	block, _ := pem.Decode(data)
	if block == nil {
			log.Fatalf("PEMデータのデコードに失敗しました: %s", filePath)
	}

	// PKCS#1, PKCS#8, EC形式の秘密鍵に対応
	var key interface{}
	key, err = x509.ParsePKCS1PrivateKey(block.Bytes)
	if err != nil {
			// PKCS#8 の場合
			key, err = x509.ParsePKCS8PrivateKey(block.Bytes)
			if err != nil {
					// EC形式の場合
					key, err = x509.ParseECPrivateKey(block.Bytes)
					if err != nil {
							log.Fatalf("秘密鍵のパースに失敗しました: %v", err)
					}
			}
	}
	return key
}

// サーバー証明書の詳細を表示する関数
func displayServerDetails(cert *x509.Certificate) {
	fmt.Printf("サーバー証明書の詳細:\n")
	fmt.Printf("有効期限: %s\n", cert.NotAfter)
	fmt.Printf("発行者: %s\n", cert.Issuer)
	fmt.Printf("コモンネーム: %s\n", cert.Subject.CommonName)
	fmt.Printf(("SANs: %s\n"), cert.DNSNames)
	fmt.Printf("パブリックキー情報: %s\n", cert.PublicKeyAlgorithm)
	fmt.Printf("署名アルゴリズム: %s\n", cert.SignatureAlgorithm)
	fmt.Printf("Key Usage: %s\n", keyUsageToString(cert.KeyUsage))
	fmt.Printf("Extended Key Usage: %s\n", extKeyUsageToString(cert.ExtKeyUsage))
}
// 中間証明書の詳細を表示する関数
func displayIntermediateDetails(cert *x509.Certificate) {
	fmt.Printf("中間証明書の詳細:\n")
	fmt.Printf("有効期限: %s\n", cert.NotAfter)
	fmt.Printf("発行者: %s\n", cert.Issuer)
	fmt.Printf("コモンネーム: %s\n", cert.Subject.CommonName)
	fmt.Printf("サブジェクト: %s\n", cert.Subject)
	fmt.Printf(("SANs: %s\n"), cert.DNSNames)
	fmt.Printf("パブリックキー情報: %s\n", cert.PublicKeyAlgorithm)
	fmt.Printf("署名アルゴリズム: %s\n", cert.SignatureAlgorithm)
	fmt.Printf("Key Usage: %s\n", keyUsageToString(cert.KeyUsage))
	fmt.Printf("Extended Key Usage: %s\n", extKeyUsageToString(cert.ExtKeyUsage))
}

// KeyUsage を文字列に変換する関数
func keyUsageToString(ku x509.KeyUsage) string {
	var usages []string
	if ku&x509.KeyUsageDigitalSignature != 0 {
			usages = append(usages, "DigitalSignature")
	}
	if ku&x509.KeyUsageContentCommitment != 0 {
			usages = append(usages, "ContentCommitment")
	}
	if ku&x509.KeyUsageKeyEncipherment != 0 {
			usages = append(usages, "KeyEncipherment")
	}
	if ku&x509.KeyUsageDataEncipherment != 0 {
			usages = append(usages, "DataEncipherment")
	}
	if ku&x509.KeyUsageKeyAgreement != 0 {
			usages = append(usages, "KeyAgreement")
	}
	if ku&x509.KeyUsageCertSign != 0 {
			usages = append(usages, "CertSign")
	}
	if ku&x509.KeyUsageCRLSign != 0 {
			usages = append(usages, "CRLSign")
	}
	if ku&x509.KeyUsageEncipherOnly != 0 {
			usages = append(usages, "EncipherOnly")
	}
	if ku&x509.KeyUsageDecipherOnly != 0 {
			usages = append(usages, "DecipherOnly")
	}
	return fmt.Sprintf("[%s]", stringJoin(usages, ", "))
}
// ExtKeyUsage を文字列に変換する関数
func extKeyUsageToString(extUsages []x509.ExtKeyUsage) string {
	var usages []string
	for _, usage := range extUsages {
			switch usage {
			case x509.ExtKeyUsageAny:
					usages = append(usages, "Any")
			case x509.ExtKeyUsageServerAuth:
					usages = append(usages, "ServerAuth")
			case x509.ExtKeyUsageClientAuth:
					usages = append(usages, "ClientAuth")
			case x509.ExtKeyUsageCodeSigning:
					usages = append(usages, "CodeSigning")
			case x509.ExtKeyUsageEmailProtection:
					usages = append(usages, "EmailProtection")
			case x509.ExtKeyUsageIPSECEndSystem:
					usages = append(usages, "IPSECEndSystem")
			case x509.ExtKeyUsageIPSECTunnel:
					usages = append(usages, "IPSECTunnel")
			case x509.ExtKeyUsageIPSECUser:
					usages = append(usages, "IPSECUser")
			case x509.ExtKeyUsageTimeStamping:
					usages = append(usages, "TimeStamping")
			case x509.ExtKeyUsageOCSPSigning:
					usages = append(usages, "OCSPSigning")
			case x509.ExtKeyUsageMicrosoftServerGatedCrypto:
					usages = append(usages, "MicrosoftServerGatedCrypto")
			case x509.ExtKeyUsageNetscapeServerGatedCrypto:
					usages = append(usages, "NetscapeServerGatedCrypto")
			default:
					usages = append(usages, "Unknown")
			}
	}
	return fmt.Sprintf("[%s]", stringJoin(usages, ", "))
}

// スライスをカンマ区切りの文字列に変換するヘルパー関数
func stringJoin(elements []string, sep string) string {
	result := ""
	for i, elem := range elements {
			if i > 0 {
					result += sep
			}
			result += elem
	}
	return result
}

File: node-cert-tools/verify-cert.ts
#!/usr/bin/env node

import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";

// メイン関数
function main(): void {
	// 引数が渡されているか確認
	if (process.argv.length < 3) {
		console.log("Usage: node verify-cert.ts <DOMAIN>");
		return;
	}

	// ドメイン名を取得
	const domain: string = process.argv[2];
	console.log(`Verifying certificate for domain: ${domain}`);

	// fqdn を宣言して値を代入
	const fqdn: string = domain;
	const localPath: string = "../certificates/";

	// ファイルパスを指定
	const certFile: string = path.join(localPath, fqdn, "cert.pem"); // サーバー証明書
	const chainFile: string = path.join(localPath, fqdn, "chain.pem"); // 中間証明書
	const keyFile: string = path.join(localPath, fqdn, "privkey.pem"); // サーバー証明書の秘密鍵

	// 証明書と鍵を読み込む
	const serverCert = loadCertificate(certFile);
	const intermediateCert = loadCertificate(chainFile);
	const privateKey = loadPrivateKey(keyFile);

	// サーバー証明書の詳細を表示
	displayServerDetails(serverCert);
	// 中間証明書の詳細を表示
	displayIntermediateDetails(intermediateCert);

	// サーバー証明書と中間証明書の整合性を確認
	try {
		verifySignature(serverCert, intermediateCert);
		console.log("サーバー証明書と中間証明書の整合性が確認されました");
	} catch (err: unknown) {
		if (err instanceof Error) {
			console.error(
				`サーバー証明書と中間証明書の整合性が確認できません: ${err.message}`,
			);
		} else {
			console.error(
				`サーバー証明書と中間証明書の整合性が確認できません: ${String(err)}`,
			);
		}
		process.exit(1);
	}

	// サーバー証明書と秘密鍵の整合性を確認
	try {
		verifyKeyPair(serverCert, privateKey);
		console.log("サーバー証明書と秘密鍵の整合性が確認されました");
	} catch (err: unknown) {
		if (err instanceof Error) {
			console.error(
				`サーバー証明書と秘密鍵の整合性が確認できません: ${err.message}`,
			);
		} else {
			console.error(
				`サーバー証明書と秘密鍵の整合性が確認できません: ${String(err)}`,
			);
		}
		process.exit(1);
	}
}

// 証明書を読み込む関数
function loadCertificate(filePath: string): crypto.X509Certificate {
	try {
		const data = fs.readFileSync(filePath, "utf8");
		const cert = new crypto.X509Certificate(data);
		return cert;
	} catch (err: unknown) {
		if (err instanceof Error) {
			console.error(`証明書の読み込みに失敗しました: ${err.message}`);
		} else {
			console.error(`証明書の読み込みに失敗しました: ${String(err)}`);
		}
		process.exit(1);
	}
}

// 秘密鍵を読み込む関数
function loadPrivateKey(filePath: string): crypto.KeyObject {
	try {
		const data = fs.readFileSync(filePath, "utf8");
		return crypto.createPrivateKey(data);
	} catch (err: unknown) {
		if (err instanceof Error) {
			console.error(`秘密鍵の読み込みに失敗しました: ${err.message}`);
		} else {
			console.error(`秘密鍵の読み込みに失敗しました: ${String(err)}`);
		}
		process.exit(1);
	}
}

// サーバー証明書の詳細を表示する関数
function displayServerDetails(cert: crypto.X509Certificate): void {
	console.log("サーバー証明書の詳細:");
	console.log(`有効期限: ${cert.validTo}`);
	console.log(`発行者: ${cert.issuer}`);
	console.log(`コモンネーム: ${getCommonName(cert.subject)}`);
	console.log(`SANs: ${cert.subjectAltName || "[]"}`);
	// @ts-ignore
	console.log(`パブリックキー情報: ${cert.publicKey.asymmetricKeyType}`);
	// @ts-ignore
	console.log(`署名アルゴリズム: ${cert.sigalg}`);
	// Node.jsの標準ライブラリではKeyUsageとExtKeyUsageの詳細表示は制限があります
}

// 中間証明書の詳細を表示する関数
function displayIntermediateDetails(cert: crypto.X509Certificate): void {
	console.log("中間証明書の詳細:");
	console.log(`有効期限: ${cert.validTo}`);
	console.log(`発行者: ${cert.issuer}`);
	console.log(`コモンネーム: ${getCommonName(cert.subject)}`);
	console.log(`サブジェクト: ${cert.subject}`);
	console.log(`SANs: ${cert.subjectAltName || "[]"}`);
	// @ts-ignore
	console.log(`パブリックキー情報: ${cert.publicKey.asymmetricKeyType}`);
	// @ts-ignore
	console.log(`署名アルゴリズム: ${cert.sigalg}`);
	// Node.jsの標準ライブラリではKeyUsageとExtKeyUsageの詳細表示は制限があります
}

// サブジェクトからコモンネームを抽出する関数
function getCommonName(subject: string): string {
	const cnMatch = subject.match(/CN=([^,]+)/);
	return cnMatch ? cnMatch[1] : "N/A";
}

// サーバー証明書と中間証明書の整合性を確認する関数
function verifySignature(
	serverCert: crypto.X509Certificate,
	intermediateCert: crypto.X509Certificate,
): void {
	// @ts-ignore
	const verified = serverCert.verify(intermediateCert.publicKey);
	if (!verified) {
		throw new Error("署名の検証に失敗しました");
	}
}

// サーバー証明書と秘密鍵の整合性を確認する関数
function verifyKeyPair(
	cert: crypto.X509Certificate,
	privateKey: crypto.KeyObject,
): void {
	// @ts-ignore
	const publicKey = cert.publicKey;
	const testData = Buffer.from("test");
	try {
		const signature = crypto.sign("sha256", testData, privateKey);
		const verified = crypto.verify("sha256", testData, publicKey, signature);
		if (!verified) {
			throw new Error("鍵ペアの検証に失敗しました");
		}
	} catch (err: unknown) {
		if (err instanceof Error) {
			throw new Error(`鍵ペアの検証中にエラーが発生しました: ${err.message}`);
		}
		throw new Error(`鍵ペアの検証中にエラーが発生しました: ${String(err)}`);
	}
}

// メイン関数を実行
main();
